### PR TITLE
Lens Flare various fixes

### PR DESCRIPTION
**Pull request by sebastienlagarde**
### Purpose of this PR
Various small fixes a features request.

Fixes:
- Change count of Element Count + Undo (produces error)
- https://fogbugz.unity3d.com/f/cases/1346894/
- Fix color support for: Thumbnail for Element with Distribution "Random" and Length as 0 (only case supported)
- Remake "radialScreenAttenuationCurve" visible
- Reorder LensFlare position on PostProcess to have the same behavior on HDRP & URP
- Fix rotation: https://fogbugz.unity3d.com/f/cases/1348570/
- Fix missing Uniform Angle for Uniform distribution (support thumbnail too)

Simplification/Optim:
- Reduce memory footprint don't need anymore _FlareData5 (on code side, no visual impact).

---
### Testing status
Bug fixes tested
Tested URP & HDRP side by side to see same results with various parameter randomization.
Now the difference would be, the LensFlare would be influenced the UberShader which implies:
* Bloom
* Color Grading
* Lens Distortion
* Chromatic Aberation
* Vigneting

---
### Comments to reviewers
Make sure the reorder of PostProcess don't have missing edges cases {URP + HDRP}.

Fixes:
https://fogbugz.unity3d.com/f/cases/1346894/

Yamato:
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/HDRP%252Flens_flare_various_fixes3-bis/.yamato%252Fall-hdrp.yml%2523PR_HDRP_2021.2/7654618/job

### CODE REVIEW

**Review by skhiat — APPROVED** [review]

**Review by RSlysz — APPROVED** [review]
lgtm

File: com.unity.render-pipelines.core/Documentation~/Common/lens-flare-data-driven-asset.md [hunk 1]
```diff
@@ -134,6 +136,7 @@ The Lens Flare Element asset has the following properties:
 | **Property**    | **Description**                                              |
 | --------------- | ------------------------------------------------------------ |
 | Colors | The range of colors that this asset applies to the lens flares. |
+| Rotation | Specify an angle (in degree) which will be used for each element incrementaly. |
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.core/Documentation~/Common/lens-flare-data-driven-asset.md:139]: | Rotation | The angle of rotation (in degrees) applied to each element incrementally. |

File: com.unity.render-pipelines.core/Documentation~/Common/lens-flare-data-driven-asset.md [hunk 2]
```diff
@@ -145,7 +148,8 @@ The Lens Flare Element asset has the following properties:
 | ---------------- | ------------------------------------------------------------ |
 | Colors           | The range of colors that this asset applies to the lens flares. You can use the **Position Spacing** curve to determine how this range affects each lens flare. |
 | Position Spacing | Adjust this curve to change the placement of the lens flare elements in the **Lens Spread**. |
-| Scale Variation  | Adjust this curve to control the size range of the lens flare elements. |
+| Rotation | Uniform angle (in degree) used with multiple element enabled with Curve distribution. |
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.core/Documentation~/Common/lens-flare-data-driven-asset.md:151]: | Rotation | The uniform angle of rotation (in degrees) applied to each element distributed along the curve. |

Comment (Vic-Cooper) [reply com.unity.render-pipelines.core/Documentation~/Common/lens-flare-data-driven-asset.md:151]: Please update the screenshot to be accurate to these property changes.

File: com.unity.render-pipelines.high-definition/Documentation~/lens-flare-data-driven-asset.md [hunk 3]
```diff
@@ -134,6 +136,7 @@ The Lens Flare Element asset has the following properties:
 | **Property**    | **Description**                                              |
 | --------------- | ------------------------------------------------------------ |
 | Colors | The range of colors that this asset applies to the lens flares. |
+| Rotation | Specify an angle (in degree) which will be used for each element incrementaly. |
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/lens-flare-data-driven-asset.md:139]: As above: 
| Rotation | The angle of rotation (in degrees) applied to each element incrementally. |

File: com.unity.render-pipelines.high-definition/Documentation~/lens-flare-data-driven-asset.md [hunk 4]
```diff
@@ -145,7 +148,8 @@ The Lens Flare Element asset has the following properties:
 | ---------------- | ------------------------------------------------------------ |
 | Colors           | The range of colors that this asset applies to the lens flares. You can use the **Position Spacing** curve to determine how this range affects each lens flare. |
 | Position Spacing | Adjust this curve to change the placement of the lens flare elements in the **Lens Spread**. |
-| Scale Variation  | Adjust this curve to control the size range of the lens flare elements. |
+| Rotation | Uniform angle (in degree) used with multiple element enabled with Curve distribution. |
```
Comment (Vic-Cooper) [inline com.unity.render-pipelines.high-definition/Documentation~/lens-flare-data-driven-asset.md:151]: As above: 

Please update the screenshot to be accurate to these property changes.

| Rotation | The uniform angle of rotation (in degrees) applied to each element distributed along the curve. |

**Review by Vic-Cooper — CHANGES_REQUESTED** [review]
Suggested changes to property descriptions. Also, they require new screenshots.

**Review by Vic-Cooper — APPROVED** [review]
All good :)

**Review by RemyUnity — COMMENTED** [review]
I tested the mentioned points and it seems good to me.
Only one thing remaining : I had a issue after changing my project packages to point to this branch where the flares were rendering all their elements at the exact position of the flare object. But after reloading the project (or just looking away ? ...) the issue disappeared and never cam back. I wasn't either able to reproduce it in other conditions.
@hdb-unity is able to reproduce it more consistently, and here is a video capture of the issue from him : https://drive.google.com/file/d/1a6TcpEUxsrp5WVVwrXScvmaSxxUTgZxI/view

This is concerning but without proper repro steps it is hard to evaluate the impact.

**Review by SeanPuller — APPROVED** [review]

**Review by skhiat — APPROVED** [review]

**Review by erikabar — APPROVED** [review]
Verified issue 1346894, also tested Lens Flares with all Post Processing effects

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs [hunk 5]
```diff
@@ -241,20 +242,21 @@ static public Vector4 GetFlareData0(Vector2 screenPos, Vector2 translationScale,
...
-            if (!autoRotate)
-            {
-                rotation = Mathf.Abs(rotation) < 1e-4f ? -360.0f : -rotation;
-            }
-            else
+            //if (!autoRotate)
+            //{
+            //    //rotation = Mathf.Abs(rotation) < 1e-4f ? 360.0f : rotation;
+            //}
+            //else
```
Comment (phi-lira) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs:249]: Commented code should be removed. Not blocking issue.

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs [hunk 6]
```diff
@@ -241,20 +242,21 @@ static public Vector4 GetFlareData0(Vector2 screenPos, Vector2 translationScale,
...
+            if (autoRotate)
             {
                 Vector2 pos = (rayOff.normalized * vLocalScreenRatio) * translationScale;
-                rotation -= Mathf.Rad2Deg * Mathf.Atan2(pos.y, pos.x);
+                rotation += -Mathf.Rad2Deg * Mathf.Atan2(pos.y, pos.x);
```
Comment (phi-lira) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs:253]: Not sure I follow this change.

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs [hunk 7]
```diff
@@ -690,10 +708,12 @@ float RandomRange(float min, float max)
                                 float sizeCurveValue = element.scaleCurve.length > 0 ? element.scaleCurve.Evaluate(timeScale) : 1.0f;
                                 localSize *= sizeCurveValue;
 
-                                Vector4 flareData0 = GetFlareData0(screenPos, element.translationScale, vScreenRatio, element.rotation, localPos, element.angularOffset, element.positionOffset, element.autoRotate);
+                                float angleFromCurve = element.uniformAngleCurve.Evaluate(timeScale);
+                                //float angleFromCurve = element.uniformAngleCurve.Evaluate(timeScale) * (180.0f - (180.0f / element.count));
```
Comment (phi-lira) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareCommonSRP.cs:712]: comments should be removed

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareDataSRP.cs [hunk 8]
```diff
@@ -294,6 +295,12 @@ public float sdfRoundness
 
         /// <summary> Inverse the gradient direction. </summary>
         public bool inverseSDF;
+
+        /// <summary> Uniform angle used with multiple element enabled with Uniform distribution. </summary>
```
Comment (phi-lira) [inline com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareDataSRP.cs:299]:
```suggestion
        /// <summary> Uniform angle in degrees used with multiple element enabled with Uniform distribution. </summary>
```

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs [hunk 9]
```diff
@@ -915,12 +915,24 @@ static float GetLensFlareLightAttenuation(Light light, Camera cam, Vector3 wo)
 
         void DoLensFlareDatadriven(Camera camera, CommandBuffer cmd, RenderTargetIdentifier source, bool usePanini, float paniniDistance, float paniniCropToFit)
         {
+            var nonJitteredCameraProj = camera.projectionMatrix;
+            var cameraProj = nonJitteredCameraProj;
+            var gpuProj = GL.GetGPUProjectionMatrix(cameraProj, false);
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs:920]: These vars are not used anywhere?

**Review by phi-lira — APPROVED** [review]
Minor comments but nothing blocking.

File: com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs (hunk 9 above)
Comment (skhiat) [reply com.unity.render-pipelines.universal/Runtime/Passes/PostProcessPass.cs]: some used to produce:
gpuVP used bellow.
I'll remove the unused one.

**Review by skhiat — APPROVED** [review]

File: com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareDataSRP.cs (hunk 8 above)
Comment (skhiat) [reply com.unity.render-pipelines.core/Runtime/PostProcessing/LensFlareDataSRP.cs:299]: That's not in degrees

**Pull request merged**